Source/Core/Common/Arm64Emitter.h

```diff
@@ -224,9 +224,65 @@ enum ARM64Reg
   WZR = WSP,
   ZR = SP,
 
-  INVALID_REG = 0xFFFFFFFF
+  INVALID_REG = -1,
 };
 
+constexpr int operator&(const ARM64Reg& reg, const int mask)
+{
+  return static_cast<int>(reg) & mask;
+}
+constexpr int operator&(const int mask, const ARM64Reg& reg)
+{
+  return static_cast<int>(reg) & mask;
+}
+constexpr int operator&(const ARM64Reg& lhs, const ARM64Reg& rhs)
+{
+  return static_cast<int>(lhs) & static_cast<int>(rhs);
+}
+constexpr int operator|(const ARM64Reg& reg, const int mask)
+{
+  return static_cast<int>(reg) | mask;
+}
+constexpr int operator|(const int mask, const ARM64Reg& reg)
+{
+  return static_cast<int>(reg) | mask;
+}
+constexpr int operator|(const ARM64Reg& lhs, const ARM64Reg& rhs)
+{
+  return static_cast<int>(lhs) | static_cast<int>(rhs);
+}
+constexpr ARM64Reg operator+(const ARM64Reg& reg, const int addend)
+{
+  return static_cast<ARM64Reg>(static_cast<int>(reg) + addend);
+}
+constexpr ARM64Reg operator+(const int addend, const ARM64Reg& reg)
+{
+  return static_cast<ARM64Reg>(static_cast<int>(reg) + addend);
+}
+constexpr ARM64Reg operator+(const ARM64Reg& lhs, const ARM64Reg& rhs)
+{
+  return static_cast<ARM64Reg>(static_cast<int>(lhs) + static_cast<int>(rhs));
+}
+constexpr ARM64Reg operator-(const ARM64Reg& reg, const int addend)
+{
+  return static_cast<ARM64Reg>(static_cast<int>(reg) - addend);
+}
+constexpr ARM64Reg operator-(const int addend, const ARM64Reg& reg)
+{
+  return static_cast<ARM64Reg>(addend - static_cast<int>(reg));
+}
+constexpr ARM64Reg operator-(const ARM64Reg& lhs, const ARM64Reg& rhs)
+{
+  return static_cast<ARM64Reg>(static_cast<int>(lhs) - static_cast<int>(rhs));
+}
+constexpr int operator<<(const ARM64Reg& reg, const int distance)
+{
+  return static_cast<int>(reg) << distance;
+}
+constexpr int operator>>(const ARM64Reg& reg, const int distance)
+{
+  return static_cast<int>(reg) >> distance;
+}
```

Comment:
I don't like the inconsistency of returning an int for the bitwise and shift operators and ARM64Reg for + and -, but the point of these operators is to reduce line noise by getting rid of a bunch of casts, and those return types reflect the expected type in most of those usages.

Reply:
I think this is reasonable for the most part, but why would someone want to add an (or subtract) an ARM64Reg with an ARM64Reg? ARM64Reg and int makes sense to me, but not two ARM64Reg.

Reply:
Adding two ARM64Regs is done in MOVI2FDUP and EncodeRegToSingle.  I don't think anything actually calls MOVI2FDUP so we could remove it, and if we made DecodeReg return an int (see other comment for more on the subject) then EncodeRegToSingle wouldn't need to add two ARM64Regs either.

Subtracting two registers was only done in the context of statements like fprs_in_use[VD - Q0], but since I'm replacing those anyway I suppose I could get rid of the subtraction operators.